graspy/align/seedless_procrustes.py

```diff
@@ -0,0 +1,328 @@
+import numpy as np
+from sklearn.utils import check_array
+
+from .base import BaseAlign
+from .sign_flips import SignFlips
+from .orthogonal_procrustes import OrthogonalProcrustes
+
+
+class SeedlessProcrustes(BaseAlign):
+    """
+    Implements an algorithm that matches two datasets using an orthogonal
+    matrix. Unlike OrthogonalProcrustes, this does not use a matching between
+    entries. It can even be used in the settings when the two datasets do not
+    have the same number of entries.
+
+    In graph setting, it is used to align the embeddings of two different
+    graphs, when it requires some simultaneous inference task, for example,
+    inside of the test for the equivalence of the latent distributions.
+
+    In essence, it aims to simultaneously obtain a, not necessarily 1-to-1,
+    correspondance between the vertices of the two data sets, and the
+    orthogonal matrix alignment matrix. It does via a sequence of alternating
+    steps, similarly to a regular expectation-maximization procedures. The step
+    of obtaining an assignment ("E-step") is done by solving an optimal
+    transport problem via Sinkhorn algorithm, whereas obtaining an orthogonal
+    alignment matrix ("M-step") is done via regular orthogonal procurstes.
+
+    Parameters
+    ----------
+        optimal_transport_lambda : float (default=0.1), optional
+            Regularization term of the Sinkhorn optimal transport algorithm.
+
+        optimal_transport_eps : float (default=0.01), optional
+            Tolerance parameter for the each Sinkhorn optimal transport
+            algorithm. I.e. tolerance for each "E-step".
+
+        optimal_transport_num_reps : int (default=1000), optional
+            Number of repetitions in each iteration of the iterative optimal
+            transport problem. I.e. maximum number of repetitions in each
+            "E-step".
+
+        iterative_eps : float (default=0), optional
+            Tolerance of the iterative optimal transport problem. I.e.
+            tolerance of the whole "EM" algorithm. Zero by default, which means
+            that the algorithm performs all `iterative_num_reps` number of
+            steps.
+
+        iterative_num_reps : int (default=100), optional
+            Number of reps in each iteration of the iterative optimal transport
+            problem. I.e. maxumum number of total iterations the whole "EM"
+            algorithm.
+
+        init : string, {"2d" (default), "sign_flips", "custom"}, optional
+
+            - "2d"
+                uses 2^d different initiazlizations, where d is the dimension.
+                specifically, uses all possible matrices with all entries real
+                and diagonal entries having magnitude 1 and 0s everywehre else.
+                for example, for d=2, tries [[1, 0], [0, 1]], [[1, 0], [0,
+                -1]], [[-1, 0], [0, 1]], and [[-1, 0], [0, -1]]. picks the best
+                one based on the value of the objective function.
```

Comment:
unclear

Reply:
more generally, this is a beast, to code and to explain. at some point we should talk about which initialization schemes are "proven" to be better than which others. right now I feel like we don't really know? so I am wondering which initialization schemes to include at all. for instance, when are 2^d sign flips better than 2^d random orthogonals? Or when should we use 2^d run to completion vs 2^d and pick one with best initial objective? I feel like there are so many ways one could do this that im struggling to figure out which it makes the most sense to include in the class (right now, at least)

Reply:
of note, this was the biggest hang up for ali's graph matching PR to scipy. since there are a bazillion ways to initialize. needed to strike some balance between simplicity and performance/flexibility. so it is just on my mind a lot right now

Reply:
well, custom seems to be a no-brainer for me, because having it allows to use any initialization schema one wants explicitly. i believe there is a reason why we check 2^d sign flip matrices as opposed 2^d random orthogonal matices. i think this stems from the 2 different non-identifiabilities that Agterberg's paper is about. and one kind of non-identifiability can only be a sign flip, whereas the other is block-orthogonal matricecs, where these blocks occur if two eigenvalues are close to each other. 

if you scroll down here: https://nbviewer.jupyter.org/url/alyakin314.github.io/assets/seedless_3.ipynb?flush_cache=true , i show the orthogonal matrix initialized and final. note that how dimensions (2 and 3), and (5 and 6) have non-zero in their mixing entry, but all others are essentially pure, so it seems like 2-d full restarts is at least admissiable. now, of u had computational power to do X restarts, where X >> 2^d, then starting with orthogonal matrices would seem like a way to go, because eventually you'd land in the neighbourhood of the global min.

Reply:
and sign-flip is sort of a semi-measure of not doing 2^d of them, which obviously can fail sometimes, as we know.

Reply:
by the way, in my experiments, i am yet to stumble upon a case where 2^d restarts fail. they sometimes can land in the local minimas when  there is a true clear global minima known to me ( for this - see: https://nbviewer.jupyter.org/url/alyakin314.github.io/assets/seedless_1.ipynb?flush_cache=true ), but the result is still "good".

Reply:
oops, forgot to rephrase this

Reply:
see now